### PR TITLE
ruby: build native Darwin gems using rake-compiler-dock

### PR DESCRIPTION
**Pull request by flavorjones**
Summary of changes:
- add `x86_64-darwin` ~and `arm64-darwin`~ native platforms
  - (Update 2021-04-16: Note that `arm64-darwin` support has been moved to a separate PR at #25992)
- bump dev dependency on rake-compiler-dock to `~> 1.1` and stop setting `no_native` on the extension task
- add shared object files to the Rake `CLEAN` file list, and remove those files between native gem builds
- use `bundle exec` in the RCD container

The first five commits are small prefactors or cleanup; the interesting bit is in the final commit.

Using rake-compiler-dock ("RCD") for these platforms unifies the Darwin native gem build process with the Linux native gems, which should help avoid inconsistencies in packaging that result in issues like the missing Ruby 3.0 binaries in #25060.

Please note that this change leaves the "universal-darwin" platform native gem untouched, but provides a path forward if the project ever decides to drop "universal" binary support.

Related to:

- #25429
- #25755
- #25756

### CODE REVIEW

File: src/ruby/ext/grpc/extconf.rb
Comment (apolcyn) [inline src/ruby/ext/grpc/extconf.rb:50]: Unless I'm missing something, all of these macos-specific options are still needed for people building the gem from source on macos hosts.

**Comment by casperisfine** [comment]
> in issues
> like #25060,

I suppose you meant o link my comment? https://github.com/grpc/grpc/issues/25060#issuecomment-804132905 (darwin packages not containing the 3.0 `.bundle`)

File: src/ruby/ext/grpc/extconf.rb
Comment (flavorjones) [reply src/ruby/ext/grpc/extconf.rb:50]: :facepalm: Let me fix that.

**Comment by flavorjones** [comment]
This PR now is purely additive: it adds the capability to generate `x86_64-darwin` and `arm64-darwin` native gems. The capability still exists for building the `universal-darwin` native gem ... err, natively ... on a Darwin system; as well as the ability for all supported platforms to compile the gem from source. The PR description has been updated with more details.

**Comment by apolcyn** [comment]
I've kicked off a package build in https://fusion2.corp.google.com/invocations/b10cd119-1278-4530-888d-5d70120ef3b3/targets - after that's done, we should be able to download the new macos packages and test.

**Comment by flavorjones** [comment]
@apolcyn The docker images `grpctesting/rake_x86_64-darwin` and `grpctesting/rake_arm64-darwin` need to be created (by `push_testing_images.sh`). I looked but couldn't see a place where this is automated -- is it possible that this is a manual step?

The failure messages I saw were all variations of:
```
Unable to find image 'grpctesting/rake_x86_64-darwin:e738cc1e6f4dd11445590a3ab81022fd70282395' locally
docker: Error response from daemon: pull access denied for grpctesting/rake_x86_64-darwin, repository does not exist or may require 'docker login'.
```

**Comment by apolcyn** [comment]
> @apolcyn The docker images grpctesting/rake_x86_64-darwin and grpctesting/rake_arm64-darwin need to be created (by push_testing_images.sh). I looked but couldn't see a place where this is automated -- is it possible that this is a manual step?

Yes, this is a manual steo done via the `tools/dockerfile/push_testing_images.sh` script. One needs to have write access to the docker repo though to run that script -- I've ran it, so the docker images should be there now.

Kicked off another package build in https://fusion2.corp.google.com/invocations/b6c8f76a-6d10-4815-9f28-9ce8c5e4e5db/targets

**Comment by flavorjones** [comment]
Looks like I needed to also update `templates/grpc.gemspec.template` with the `rake-compiler-dock` dependency upgrade. I've pushed that change now.

**Comment by flavorjones** [comment]
@apolcyn Would you mind kicking CI again to see if it's green after yesterday's additional fix to the gemspec template?

File: Rakefile
Comment (apolcyn) [inline Rakefile:147]: I believe that we we run `rake gem:native` on macos hosts (for the mac package build), this will also build the `'x86_64-darwin` packages on macos - so I want to say we're building this packages twice: once on linux via docker and again on macos.

Since this set of cross-compilation rake tasks is only needed for cross-compilation from linux, can we run this conditionally, only under linux?

I think that will take care of the macos package build timeout that happened in the last package build link posted.

Comment (apolcyn) [reply Rakefile:147]: Whoops! Sorry yes you're right. I've kicked off another package build:  https://fusion2.corp.google.com/invocations/684801ca-5bca-4ae1-90b7-8f296300ef00 (the previous timeout may just be a flake):

Comment (flavorjones) [reply Rakefile:147]: Thanks! I can't access that link, but please let me know if it goes green?

Comment (apolcyn) [reply Rakefile:147]: https://fusion2.corp.google.com/invocations/684801ca-5bca-4ae1-90b7-8f296300ef00 is green!

**Comment by flavorjones** [comment]
I don't think the remaining failure is related (it appears to be an issue building [csharp artifacts](https://source.cloud.google.com/results/invocations/f15e90bc-c248-440e-8044-42df4b839683/targets/github%2Fgrpc/tests)).

**Comment by apolcyn** [comment]
To update, I've tested one of the packages in https://fusion2.corp.google.com/invocations/684801ca-5bca-4ae1-90b7-8f296300ef00 manually and it looks good.

I'm prioritizing getting https://github.com/grpc/grpc/pull/25429 in first though, in order to add ruby 3.0 support without risking regression for older ruby platform support (since binary packages for older platforms isn't affected by that PR). My plan is to try to get this PR in afterwards to subsequently simply things.

**Comment by apolcyn** [comment]
Circling back on this, in general I think this looks good, and it will allow us to get rid of a lot of the pain associated with building native macos packages on macos hosts - e.g. this will allow us to eventually get rid of https://github.com/grpc/grpc/blob/c4b77ca0fe8a27a2f57567e94b81130d257bd54f/tools/distrib/build_ruby_environment_macos.sh#L1 entirely.

The main issue is just that while we are adding an arm64 macos binary package, grpc doesn't have arm testing yet. I'd like to get thoughts from @jtattermusch and @nicolasnoble about this. Also note that another possibility could be to remove the binary arm packages from this PR, but keep everything else.

**Comment by apolcyn** [comment]
@flavorjones sorry can we actually remove the addition of arm packages from this PR, and, if you'd like, open a second PR that adds arm packages so that the arm addition can be considered separately from the overall build system changes?

**Comment by flavorjones** [comment]
@apolcyn On it.

**Comment by flavorjones** [comment]
@apolcyn I've pulled out the `arm64` config from this PR and rebased it onto `master`.

I have a second branch ready to go with a PR to add `arm64` config back in whenever you're ready for it.

**Comment by apolcyn** [comment]
@flavorjones not sure if you can see the log, but the failed "Artifact build linux" package build job failed with this at the bottom of the log:

```
1 warning generated.
linking shared-object grpc/grpc_c.bundle
ld: illegal thread local variable reference to regular symbol __ZN4absl12lts_2021032413base_internal19thread_identity_ptrE for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [Makefile:264: grpc_c.bundle] Error 1
rake aborted!
Command failed with status (2): [/usr/bin/make...]
/usr/local/rvm/gems/ruby-2.5.8/gems/rake-compiler-1.1.1/lib/rake/extensiontask.rb:167:in `block (2 levels) in define_compile_tasks'
/usr/local/rvm/gems/ruby-2.5.8/gems/rake-compiler-1.1.1/lib/rake/extensiontask.rb:166:in `block in define_compile_tasks'
/usr/local/rvm/gems/ruby-2.5.8/gems/rake-13.0.3/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.5.8/bin/ruby_executable_hooks:24:in `eval'
/usr/local/rvm/gems/ruby-2.5.8/bin/ruby_executable_hooks:24:in `<main>'
Tasks: TOP => native:x86_64-darwin => native:grpc:x86_64-darwin => tmp/x86_64-darwin/stage/src/ruby/lib/grpc/3.0/grpc_c.bundle => copy:grpc_c:x86_64-darwin:3.0.0 => tmp/x86_64-darwin/grpc_c/3.0.0/grpc_c.bundle
(See full trace by running task with --trace)
rake aborted!
Command failed with status (1): [docker run -v /tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64:/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64 -e UID\=1000 -e GID\=1000 -e USER\=kbuilder -e GROUP\=kbuilder -e GEM_PRIVATE_KEY_PASSPHRASE -e ftp_proxy -e http_proxy -e https_proxy -e RCD_HOST_RUBY_PLATFORM\=x86_64-linux -e RCD_HOST_RUBY_VERSION\=2.4.1 -e RCD_IMAGE\=grpctesting/rake_x86_64-darwin:e738cc1e6f4dd11445590a3ab81022fd70282395 -w /tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64 --rm -i grpctesting/rake_x86_64-darwin:e738cc1e6f4dd11445590a3ab81022fd70282395 runas sigfw bash -c gem\ update\ --system\ --no-document\ \&\&\ \ \ \ \ \ \ \ \ bundle\ \&\&\ \ \ \ \ \ \ \ \ bundle\ exec\ rake\ clean\ \&\&\ \ \ \ \ \ \ \ \ bundle\ exec\ rake\ native:x86_64-darwin\ pkg/grpc-1.38.0.dev-x86_64-darwin.gem\ pkg/grpc-1.38.0.dev.gem\ \ \ \ \ \ \ \ \ \ \ RUBY_CC_VERSION\=3.0.0:2.7.0:2.6.0:2.5.0:2.4.0\ \ \ \ \ \ \ \ \ \ \ V\=0\ \ \ \ \ \ \ \ \ \ \ GRPC_CONFIG\=opt'
']
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-compiler-dock-1.1.0/lib/rake_compiler_dock/starter.rb:96:in `block in exec'
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-compiler-dock-1.1.0/lib/rake_compiler_dock/starter.rb:46:in `each'
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-compiler-dock-1.1.0/lib/rake_compiler_dock/starter.rb:46:in `exec'
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-compiler-dock-1.1.0/lib/rake_compiler_dock/starter.rb:14:in `sh'
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-compiler-dock-1.1.0/lib/rake_compiler_dock.rb:46:in `sh'
tools/distrib/rake_compiler_docker_image.rb:36:in `run_rake_compiler'
/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/Rakefile:148:in `block (2 levels) in <top (required)>'
/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/Rakefile:147:in `each'
/tmpfs/src/github/grpc/workspace_ruby_native_gem_linux_x64/Rakefile:147:in `block in <top (required)>'
/usr/local/rvm/gems/ruby-2.4.1/gems/rake-13.0.3/exe/rake:27:in `<top (required)>'
/usr/local/rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `eval'
/usr/local/rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => gem:native
(See full trace by running task with --trace)

2021-04-16 13:38:28,705 FAILED: build_artifact.ruby_native_gem_linux_x64 [ret=1, pid=5439, time=2267.9sec]
2021-04-16 13:38:28,769 FAILED: Failed to build targets.
```

**Comment by flavorjones** [comment]
@apolcyn Thanks for the poke. I've reproduced this in my development environment and will dig into it.

**Comment by flavorjones** [comment]
A git bisect reveals that the error we're seeing:

```
ld: illegal thread local variable reference to regular symbol __ZN4absl12lts_2021032413base_internal19thread_identity_ptrE for architecture x86_64
```

was introduced in cb9889d35d29b29de8192d1c8431f93e7dc8f5b9 related to #25841, so I'm ccing @veblush who might be able to help us understand why this is happening when we cross-compile for Darwin.

**Comment by flavorjones** [comment]
A git bisect reveals that the error we're seeing:

```
ld: illegal thread local variable reference to regular symbol __ZN4absl12lts_2021032413base_internal19thread_identity_ptrE for architecture x86_64
```

was introduced in cb9889d35d29b29de8192d1c8431f93e7dc8f5b9 related to #25841, so I'm ccing @veblush who might be able to help us understand why this is happening when we cross-compile for Darwin.

**Comment by veblush** [comment]
@flavorjones TLS support wasn't that great on OSX. https://github.com/grpc/grpc/pull/24479 might be helpful to investigate the TLS issue. What we've done to address it was basically setting the minimum version of OSX supporting TLS.

**Comment by flavorjones** [comment]
@veblush Thanks for the comment, but I will admit to not understanding the nature of this linker error or how to resolve it. If I un-define `GPR_ABSEIL_SYNC` then cross-compiling works; but other than that I honestly don't know where I would start to diagnose this or propose an alternative solution.

If anybody could provide some assistance in how you'd like me to try to resolve this problem (which is not obviously related to the point of this PR), I'd be very grateful.

**Comment by flavorjones** [comment]
@veblush @apolcyn I need help unblocking this PR. The only two options I have in my toolbox are to undefine `GPR_ABSEIL_SYNC` or close this PR, and I could use your help to discern a third path that would allow merging without breaking anything.

I think this PR is a long-term benefit to the project in that it improves the reproducibility of darwin native builds, and allows for a simple extension to support arm64 builds, but I simply don't know how to move it forward given the current state of the codebase.

**Comment by veblush** [comment]
@flavorjones I guess you can merge this if there is no test failure. https://github.com/grpc/grpc/pull/25841 was reverted for a different reason so this `GPR_ABSEIL_SYNC` issue shouldn't block you.

**Comment by flavorjones** [comment]
I've rebased onto master, if someone could kick off the tests to see if this is green after https://github.com/grpc/grpc/pull/26115, I'd greatly appreciate it!

**Comment by flavorjones** [comment]
I've rebased again. Would appreciate someone kicking CI, please.

**Comment by flavorjones** [comment]
@apolcyn CI is green, would you mind taking a look at the changes here and approving?

**Comment by apolcyn** [comment]
I'm waiting for a CI issue to be resolved so that I can kick off another full package build

**Comment by flavorjones** [comment]
@apolcyn It's been a couple of weeks, I was wondering if you were still planning on merging this?

**Comment by apolcyn** [comment]
Ok, I've kicked off another full package build with this PR in https://fusion2.corp.google.com/invocations/a3e39570-3262-48f0-aeec-5d0ec2922f46

**Comment by apolcyn** [comment]
Hi @flavorjones, can you please merge the latest master branch into your pull request branch and re-push?

The full package builds based off of your branch have been failing with:

```
++ KEYSERVER_IPV4_ADDRESS=
++ gpg --keyserver hkp:// --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
gpg: could not parse keyserver URL
```

... which is an error that's been fixed on the head of the master branch.

Sorry for the slowness in this review

**Comment by flavorjones** [comment]
@apolcyn :+1: I've rebased! Thanks again for your time.

**Comment by flavorjones** [comment]
@apolcyn I've updated this PR to work with `templates/src/ruby/ext/grpc/extconf.rb.template` which should fix the failures I was able to see from the previous run.

**Comment by flavorjones** [comment]
I've rebased again. I'd be thankful if @apolcyn or @jtattermusch would kick CI.

**Comment by jtattermusch** [comment]
Looks like the linux artifact build now creates the `grpc-1.41.0.dev-x86_64-darwin.gem`:
 
https://console.cloud.google.com/storage/browser/grpc-testing-kokoro-prod/test_result_public/prod/grpc/core/pull_request/linux/grpc_build_artifacts/27493/20210910-024653

The grpc-1.41.0.dev-universal-darwin.gem built on by the macos/grpc_build_artifacts is still being built.

@apolcyn since the tests are looking good, do you want to do the final review of this PR?

**Review by jtattermusch — APPROVED** [review]
LGTM.

**Pull request merged**
